src/components/base-component.js

```diff
@@ -0,0 +1,67 @@
+export default class BaseComponents {
+  subElements = {};
+
+  init() {
+    this.beforeRender();
+    this.render();
+    this.getSubElements();
+    this.afterRender();
+  }
+
+  beforeRender() {
+    // This method will be called before render method
+  }
+
+  get template() {
+    return ``;
+  }
+
+  render() {
+    const element = document.createElement("div");
+
+    element.innerHTML = this.template;
+
+    this.element = element.firstElementChild;
+  }
+
+  afterRender() {
+    // This method will be called after render method
+  }
+
+  getSubElements() {
+    const elements = this.element.querySelectorAll("[data-element]");
+
+    for (const subElement of elements) {
+      const name = subElement.dataset.element;
+
+      this.subElements[name] = subElement;
+    }
```

Comment:
**Scalability Issue**: Querying DOM elements in a loop with 'querySelectorAll' and then iterating over them to set properties can be highly inefficient, especially for large DOM structures or frequent operations. <br> **Fix**: Cache queried elements or use a more efficient method of DOM manipulation to reduce reflow and repaint costs. <br> **Code Suggestion**: 
 ```
 - const elements = this.element.querySelectorAll("[data-element]");
 - for (const subElement of elements) {
 -   const name = subElement.dataset.element;
 -   this.subElements[name] = subElement;
 - }
 + // Use more efficient DOM manipulation methods
 ```